# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/wr/DisplayItemCache.cpp
# Commit: aaed31bbe8f4
# Full Hash: aaed31bbe8f4e0a8892e632633c27033549eaf33
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2020-03-19 03:53:34
# Regressor Bug: 1614655
# File Overlap Count: 2
# Description:
#   Bug 1614655 - Part 2: Allow 1:n mapping of Gecko - WR display items r=jrmuizel
#   
#   This patch changes the underlying storage for WR display items in DisplayItemCache
#   from Vec<Option<CachedDisplayItem> to Vec<Vec<CachedDisplayItem>>.
#   This allows storing multiple WebRender display items for one Gecko display item.
# ==============================================================================

diff -r 0e791acac783 -r aaed31bbe8f4 gfx/layers/wr/DisplayItemCache.cpp
--- a/gfx/layers/wr/DisplayItemCache.cpp	Wed Mar 18 23:46:27 2020 +0000
+++ b/gfx/layers/wr/DisplayItemCache.cpp	Wed Mar 18 23:47:05 2020 +0000
@@ -85,7 +85,7 @@
     return Nothing();
   }
 
-  if (!aItem->CanBeReused()) {
+  if (!aItem->CanBeReused() || !aItem->CanBeCached()) {
     // Do not try to cache items that cannot be reused.
     return Nothing();
   }
@@ -129,14 +129,16 @@
     return Nothing();
   }
 
-  // Spatial id and clip id can change between display lists.
   if (!(aSpaceAndClip == slot.mSpaceAndClip)) {
-    // Mark the cache slot inactive and recache the item.
+    // Spatial id and clip id can change between display lists, if items that
+    // generate them change their order.
     slot.mOccupied = false;
-    return Nothing();
+    aItem->SetCantBeCached();
+    slotIndex = Nothing();
+  } else {
+    slot.mUsed = true;
   }
 
-  slot.mUsed = true;
   return slotIndex;
 }
 